Move helper structs/methods out of `ir_emitter.cpp`

ansley opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #59465
* #59245
* __->__ #59244

Differential Revision: [D28797511](https://our.internmc.facebook.com/intern/diff/D28797511)

Commit:
Move helper structs/methods out of `ir_emitter.cpp`

[ghstack-poisoned]

ansley commented:
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28797511).

zhxchen17 commented on torch/csrc/jit/frontend/ir_emitter.h:
```diff
@@ -13,7 +13,7 @@
 namespace jit {
 
-TORCH_API void runCleanupPasses(std::shared_ptr<Graph>& to_clean);
+TORCH_API void RunCleanupPasses(std::shared_ptr<Graph>& to_clean);
```
Why Capitalizing™ this?

ansley replied:
The other passes (e.g. `ConvertToSSA`, `CanonicalizeModifiedLoops`, `NormalizeOps`) are capitalized, so I thought we should maintain that consistency. I can switch it back, though

zhxchen17 replied:
I see! One small issue is that it's hard to tell whether we are running a plain function or constructing an object in place by looking at `PassName(...)` call, since typically we capitalize type names, although it's easier to tell it's a function call here. I guess this is not urgent :), but it'd be great if we can rename all passes to start with lower case char.

ansley replied:
I like that better; I'll rename all the passes to be in `camelCase` if they aren't objects

zhxchen17 commented on torch/csrc/jit/frontend/environment.h:
```diff
@@ -0,0 +1,384 @@
+#include <torch/csrc/jit/frontend/ir_emitter_utils.cpp>
```
The fact that we have to include an implementation file usually indicates bigger problems. What we're trying to accomplish by including a cpp file instead of an h file?

ansley replied:
Good point, `ir_emitter_utils` should be a header file since it can compile independently

zhxchen17 reviewed (changes requested):
great, seems like ir_emitter.cpp will be more friendly to new learners like me 👍

Commit:
Update on "Move helper structs/methods out of `ir_emitter.cpp`"


Differential Revision: [D28797511](https://our.internmc.facebook.com/intern/diff/D28797511)

[ghstack-poisoned]

SplitInfinity commented on torch/csrc/jit/frontend/run_cleanup_passes.h:
```diff
@@ -0,0 +1,47 @@
+namespace torch {
+namespace jit {
+
+void RunCleanupPasses(std::shared_ptr<Graph>& graph) {
```
Consider putting the implementation into a `.cpp`.

SplitInfinity commented on torch/csrc/jit/frontend/type_refinement.h:
```diff
@@ -0,0 +1,177 @@
+namespace torch {
+namespace jit {
+
+struct Refinement {
```
Consider putting all these implementations into a `.cpp`

SplitInfinity commented on torch/csrc/jit/frontend/environment.h:
```diff
@@ -0,0 +1,384 @@
+// prim::Print(%a)
+
+struct Environment {
+  Environment(
```
Consider putting the implementations in this file in a `.cpp`.

SplitInfinity reviewed (commented):
Thanks for doing this! I noticed that you did not choose to put a lot of the code you moved around into `.cpp` files. Can you discuss why?

SplitInfinity commented on torch/csrc/jit/frontend/type_refinement.h:
```diff
@@ -0,0 +1,177 @@
+namespace torch {
```
Also, it is weird that aliases like TypePtr are being used here but there are no includes. I'm guessing that everything works because this file is always included after a header that directly or indirectly includes its definition.

Commit:
Update on "Move helper structs/methods out of `ir_emitter.cpp`"


Differential Revision: [D28797511](https://our.internmc.facebook.com/intern/diff/D28797511)

[ghstack-poisoned]

ansley commented:
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28797511).

zhxchen17 commented on torch/csrc/jit/frontend/ir_emitter_utils.h:
```diff
@@ -0,0 +1,421 @@
+/* ==================================================================== */
+
+// see [setstate type]
+static TypePtr getTypeForSetStateArg(const Def& def, const Self* self) {
```
`static` will cause function definitions to be copied N times if this header is included in N different translation units (files). If we meant to share this function through the header, then consider removing `static`, otherwise if this is logically private we should move this in an implementation file.

SplitInfinity replied:
I think we can remove `static` altogether. It was used in `ir_emitter.cpp` because this function was defined within the scope of `to_ir` but was not a member function.

zhxchen17 commented on torch/csrc/jit/frontend/ir_emitter_utils.h:
```diff
@@ -0,0 +1,421 @@
+}
+
+// see [setstate type]
+static bool shouldDeriveSetStateType(
```
the same `static` issue, and below.

zhxchen17 reviewed (commented):
some side notes

Commit:
Update on "Move helper structs/methods out of `ir_emitter.cpp`"


Differential Revision: [D28797511](https://our.internmc.facebook.com/intern/diff/D28797511)

[ghstack-poisoned]

ansley commented on torch/csrc/jit/frontend/ir_emitter_utils.h:
```diff
@@ -0,0 +1,426 @@
+/*                      Program Status                      */
+/* ======================================================== */
+
+enum NoneStatus { ALWAYS, MAYBE, NEVER };
```
Good point, added #59465

SplitInfinity commented on test/cpp/jit/test_cleanup_passes.cpp:
```diff
@@ -4,6 +4,7 @@
 #include <torch/csrc/jit/ir/ir.h>
 #include <torch/csrc/jit/ir/irparser.h>
 #include <torch/csrc/jit/testing/file_check.h>
+#include <torch/csrc/jit/frontend/run_cleanup_passes.cpp>
```
Why is a cpp file being included here?

ansley replied:
See: https://github.com/pytorch/pytorch/pull/59245#discussion_r644241674

Praying Nikita will save me, but would also welcome suggestions

SplitInfinity replied:
What's the linker error you are seeing?

gmagogsfm replied:
I caught a glimpse (~2s) when discussing it with @ansley. I think the error complains about some symbol visibility. Maybe it is caused by cross-library (libtorch vs libtorch_python) access of public symbols. In that case, one common solution is to add `TORCH_API` modifier in front of declaration of that symbol, which modifies the visibility. 

If it doesn't help, please post the detailed error here so that @SplitInfinity or I can help take a look.

ansley replied:
Awesome, adding `TORCH_API` fixed it

SplitInfinity commented on torch/csrc/jit/frontend/environment.h:
```diff
@@ -1,8 +1,13 @@
+#pragma once
+
 #include <torch/csrc/jit/frontend/ir_emitter_utils.h>
```
I don't know what the best practice™ regarding this is, but I think we should explicitly include the headers for types that are used in this file in case the includes inside `ir_emitter_utils.h` change in the future.

ansley replied:
Adding the extra headers leads to some `...will be initialized after [-Wreorder]...` warnings since e.g. we initialize `struct Block` in `ir.h`. I'm assuming I still shouldn't rely on transitive inclusions, right? I've changed it in the latest push

ansley replied:
Do we have a canonical way to suppress a warning for a line we know is fine?

gmagogsfm replied:
There is: https://stackoverflow.com/a/3125889

But we need some serious justification to use warning suppression I think.

What exactly is the warning you see? And what changes did you make to code to get the warning?

SplitInfinity commented on torch/csrc/jit/frontend/ir_emitter.h:
```diff
@@ -1,21 +1,319 @@
 namespace torch {
 namespace jit {
 
-TORCH_API void runCleanupPasses(std::shared_ptr<Graph>& to_clean);
```
Do all of the functions, aliases, and types declared here need to be exposed? On the master branch, this file declares only two functions: `runCleanupPasses` and `meaningfulName`. Are all of these other declarations that have been added used by files that include this header?

If not, consider moving them into the `.cpp`.

ansley replied:
I can move the `using` declarations, etc. into the `.cpp` file, but I do need to declare some of the internal methods of `to_ir` in the header file. So, I thought it would be best to declare all the internal methods in the header file. I was under the impression that best practice was keeping the "API reference" in a single location (either the header OR the source file). Is that only applicable in certain circumstances?

SplitInfinity replied:
That's true, but most of the stuff in the header is not a part of the API of `to_ir`. Unfortunately, it is not possible to split the declaration of a class into multiple pieces (like you can do for a namespace). Move what you can, but it looks like a lot of the `emitXXX` methods use the `graph` or `environment` member of `to_ir`, so they cannot be easily converted into anonymous functions that live only in the `.cpp`. We could consider using the [`pImpl` idiom](https://en.cppreference.com/w/cpp/language/pimpl) to hide these extra details, but I think that should be in a future PR.

SplitInfinity reviewed (commented):
Thanks for addressing my comments! A handful of issues still remain.

gmagogsfm commented on torch/csrc/jit/frontend/environment.h:
```diff
@@ -0,0 +1,114 @@
+        b(b),
+        next(std::move(next)) {}
+
+  // NOLINTNEXTLINE(cppcoreguidelines-non-private-member-variables-in-classes)
+  Function& method;
+  // NOLINTNEXTLINE(cppcoreguidelines-non-private-member-variables-in-classes)
+  ResolverPtr resolver;
+  // NOLINTNEXTLINE(cppcoreguidelines-non-private-member-variables-in-classes)
+  std::unordered_map<std::string, std::function<std::string()>> error_messages;
+  // NOLINTNEXTLINE(cppcoreguidelines-non-private-member-variables-in-classes)
+  Block* b;
+
+  // NOLINTNEXTLINE(cppcoreguidelines-non-private-member-variables-in-classes)
+  std::shared_ptr<Environment> next;
```
It is generally a bad code smell if lint warnings are suppressed.

Maybe you can try following approaches to improve code quality
1) Make them private or protected depending on access pattern they need.
2) Add getter methods if access to these fields are indeed needed
3) Declare classes, that need direct access to these fields, as friend classes

ansley replied:
Not my code; I just copied this and moved it to a different file. Perhaps a follow-up PR could be deleting lint warnings? Or should I do it in this PR?

ansley replied:
I went ahead and deleted them even though it's not my code. Let's consider it part of the clean-up effort :)

Commit:
Update on "Move helper structs/methods out of `ir_emitter.cpp`"


Differential Revision: [D28797511](https://our.internmc.facebook.com/intern/diff/D28797511)

[ghstack-poisoned]

Commit:
Update on "Move helper structs/methods out of `ir_emitter.cpp`"


Differential Revision: [D28797511](https://our.internmc.facebook.com/intern/diff/D28797511)

[ghstack-poisoned]

gmagogsfm reviewed (commented):
Change looks great! It should be ready for landing after broken tests are fixed.

Commit:
Update on "Move helper structs/methods out of `ir_emitter.cpp`"


Differential Revision: [D28797511](https://our.internmc.facebook.com/intern/diff/D28797511)

[ghstack-poisoned]

ansley commented:
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28797511).

Commit:
Update on "Move helper structs/methods out of `ir_emitter.cpp`"


Differential Revision: [D28797511](https://our.internmc.facebook.com/intern/diff/D28797511)

[ghstack-poisoned]

Commit:
Update on "Move helper structs/methods out of `ir_emitter.cpp`"


Differential Revision: [D28797511](https://our.internmc.facebook.com/intern/diff/D28797511)

[ghstack-poisoned]

Commit:
Update on "Move helper structs/methods out of `ir_emitter.cpp`"


Differential Revision: [D28797511](https://our.internmc.facebook.com/intern/diff/D28797511)

[ghstack-poisoned]

Commit:
Update on "Move helper structs/methods out of `ir_emitter.cpp`"


Differential Revision: [D28797511](https://our.internmc.facebook.com/intern/diff/D28797511)

[ghstack-poisoned]

ansley commented:
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28797511).

Commit:
Update on "Move helper structs/methods out of `ir_emitter.cpp`"


Differential Revision: [D28797511](https://our.internmc.facebook.com/intern/diff/D28797511)

[ghstack-poisoned]

Commit:
Update on "Move helper structs/methods out of `ir_emitter.cpp`"


Differential Revision: [D28797511](https://our.internmc.facebook.com/intern/diff/D28797511)

[ghstack-poisoned]

Commit:
Update on "Move helper structs/methods out of `ir_emitter.cpp`"


Differential Revision: [D28797511](https://our.internmc.facebook.com/intern/diff/D28797511)

[ghstack-poisoned]

Commit:
Update on "Move helper structs/methods out of `ir_emitter.cpp`"


Differential Revision: [D28797511](https://our.internmc.facebook.com/intern/diff/D28797511)

[ghstack-poisoned]

ansley commented:
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28797511).

gmagogsfm reviewed (approved):
Looks good, didn't look too carefully into all the details because I assume they are just moved instead of functional changes.

Commit:
Update on "Move helper structs/methods out of `ir_emitter.cpp`"


Differential Revision: [D28797511](https://our.internmc.facebook.com/intern/diff/D28797511)

[ghstack-poisoned]

ansley commented:
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28797511).

Commit:
Update on "Move helper structs/methods out of `ir_emitter.cpp`"


Differential Revision: [D28797511](https://our.internmc.facebook.com/intern/diff/D28797511)

[ghstack-poisoned]

ansley commented:
@ansley has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28797511).
